File: Recoil/Recoil/TextHostView.swift
//
//  TextHostView.swift
//  Recoil
//
//  Created by Leland Richardson on 12/13/17.
//  Copyright © 2017 Leland Richardson. All rights reserved.
//

import Foundation

private protocol TextFriendlyView {}

extension TextLiteralHostView: TextFriendlyView {}

public class TextHostView: UILabel, TextFriendlyView {

  private var children: [TextFriendlyView] = []
  func updateTextIfNeeded() {
    var result = ""
    for view in children {
      if let view = view as? TextLiteralHostView {
        result += view.text ?? ""
      } else if let view = view as? TextHostView {
        view.updateTextIfNeeded()
        result += view.text ?? ""
      } else {
        //TODO: fail more gracefully
        fatalError()
      }
    }
    if result != text {
      text = result
      self.yoga.markDirty()
    }
  }

  override func insertRecoilSubview(_ view: UIView, at index: Int) {
    if let view = view as? TextLiteralHostView {
      children.insert(view, at: index)
    } else if let view = view as? TextHostView {
      children.insert(view, at: index)
    } else {
      //TODO: fail more gracefully
      fatalError()
    }
  }

  override func moveRecoilSubview(from fromIndex: Int, to toIndex: Int) {
    let view = children.remove(at: fromIndex)
    if let view = view as? TextLiteralHostView {
      children.insert(view, at: toIndex)
    } else if let view = view as? TextHostView {
      children.insert(view, at: toIndex)
    } else {
      //TODO: fail more gracefully
      fatalError()
    }
  }

  override func removeRecoilSubview(from fromIndex: Int) {
    children.remove(at: fromIndex)
  }
}

File: RecoilAndroidApp/app/src/main/java/com/airbnb/android/recoil/TextHostView.kt
package com.airbnb.android.recoil

import android.content.Context
import android.view.View
import android.widget.TextView
import com.facebook.yoga.YogaNode
import com.facebook.yoga.android.YogaLayout

class TextHostView(context: Context): TextView(context), RecoilView {
  override var yogaNode: YogaNode = YogaNode()
  private var children = mutableListOf<RecoilView>()

  init {
    yogaNode.data = this
    yogaNode.setMeasureFunction(ViewMeasureFunction())
  }

  fun updateTextIfNeeded() {
    var result = ""
    for (view in children) {
      result += when (view) {
        is TextLiteralHostView -> view.text
        is TextHostView -> {
          view.updateTextIfNeeded()
          view.text
        }
        else -> throw IllegalStateException()
      }
    }
    if (result != text) {
      text = result
      yogaNode.dirty()
    }
  }

  override fun getHostView(): View = this

  override fun getYogaNodeForView(view: View): YogaNode? = null

  override fun getRecoilSubviewAt(index: Int): RecoilView? = children[index]

  override fun insertRecoilSubview(view: RecoilView, index: Int) {
    when (view) {
      is TextLiteralHostView -> children.add(index, view)
      is TextHostView -> children.add(index, view)
      else -> throw IllegalStateException()
    }
  }

  override fun moveRecoilSubview(fromIndex: Int, toIndex: Int) {
    val el = children.removeAt(fromIndex)
    insertRecoilSubview(el, toIndex)
  }

  override fun removeRecoilSubview(fromIndex: Int) {
    children.removeAt(fromIndex)
  }

  override fun getRecoilParent(): RecoilView? = parent as? RecoilView

}
